modules/core/src/main/java/org/apache/ignite/internal/jdbc/thin/JdbcThinConnection.java

```diff
@@ -1994,4 +2070,252 @@ private void processDelay(InetSocketAddress sockAddr) {
+        }
+
+        /** {@inheritDoc} */
+        @Override public BinaryMetadata metadata0(int typeId) throws BinaryObjectException {
+            BinaryMetadata meta = cache.metadata0(typeId);
+            if (meta == null) {
+                try {
+                    JdbcBinaryTypeGetResult res = doRequest(new JdbcBinaryTypeGetRequest(typeId));
+                    meta = res.meta();
+                }
+                catch (ExecutionException | InterruptedException | ClientException | SQLException e) {
+                    throw new BinaryObjectException(e);
+                }
+            }
+
+            return meta;
+        }
+
+        /**
+         * Handle update binary schema result.
+         *
+         * @param res Result.
+         * @return {@code true} if handler was waiting for result with given request ID.
+         */
+        public boolean handleResult(JdbcUpdateBinarySchemaResult res) {
+            return handleResult(res.reqId(), res);
+        }
+
+        /**
+         * Handle binary type schema result.
+         *
+         * @param res Result.
+         * @return {@code true} if handler was waiting for result with given request ID.
+         */
+        public boolean handleResult(JdbcBinaryTypeGetResult res) {
+            return handleResult(res.reqId(), res);
+        }
+
+        /** {@inheritDoc} */
+        @Override public BinaryType metadata(int typeId, int schemaId) throws BinaryObjectException {
+            BinaryType type = metadata(typeId);
+            return type != null && ((BinaryTypeImpl)type).metadata().hasSchema(schemaId) ? type : null;
+        }
+
+        /** {@inheritDoc} */
+        @Override public Collection<BinaryType> metadata() throws BinaryObjectException {
+            return cache.metadata();
+        }
+    }
+
+    /**
+     * Jdbc channel to communicate in blocking style, regardless of whether streaming mode is enabled or not.
+     */
+    private abstract class BlockingJdbcChannel {
+        /** Request ID -> Jdbc result map. */
+        private Map<Long, CompletableFuture<JdbcResult>> results = new ConcurrentHashMap<>();
+
+        /**
+         * Handles result for specified request ID.
+         *
+         * @param reqId Request id.
+         * @param res Result.
+         */
+        boolean handleResult(long reqId, JdbcResult res) {
+            boolean handled = false;
+            CompletableFuture<JdbcResult> fut = results.remove(reqId);
+            if (fut != null) {
+                fut.complete(res);
+                handled = true;
+            }
+            return handled;
+        }
+
+        /**
+         * Do request in blocking style. It just call {@link JdbcThinConnection#sendRequest(JdbcRequest)} for
+         * non-streaming mode and creates future and waits it completion when streaming is enabled.
+         *
+         * @param req Request.
+         * @return Result for given request.
+         */
+        <R extends JdbcResult> R doRequest(
```

Comment:
but why?

Reply:
When I went through the code I felt that it is natural to examine `doRequest` before `handleResult` because in practice first is called before second. Actually, minor thing.

Reply:
ok, placed it before `handleRequest`.